File: Lectures/Lecture 3/cpp/stack.cpp
#include "linked_list.cpp"

using namespace std;

class Stack : public LinkedList {
public:
    void push(int data) {
        this->prepend(data);
    }

    /** 
     * Add item to start of the linked list
    */
    int pop() {
        int data = this->head->data;
        this->remove_first_node();
        return data;
    }

    bool is_empty() {
        return (this->size() == 0);
    }

    int peek() {
        if (this->head == NULL) {
            throw std::out_of_range("Stack is empty");
        }

        return this->head->data;
    }
    
    // Size is defined in base class, no need to redefine
};

main () {
    Stack my_stack = Stack();
    cout << "My stack: " << my_stack << endl;

    cout << boolalpha;   // Ensure we can print booleans as true/false instead of 1/0
    cout << "Empty?: " << my_stack.is_empty() << endl;
    cout << endl;

    my_stack.push(1);
    cout << "My stack: " << my_stack << endl;
    my_stack.push(2);
    cout << "My stack: " << my_stack << endl;
    cout << "Empty?: " << my_stack.is_empty() << endl;
    cout << "Size:" << my_stack.size() << endl;

    cout << "Peek: " << my_stack.peek() << endl;
    cout << endl;

    cout << "Pop: " << my_stack.pop() << endl;
    cout << "My stack: " << my_stack << endl;
    cout << "Pop: " << my_stack.pop() << endl;
    cout << "My stack: " << my_stack << endl;
    cout << "Peek:" << my_stack.peek() << endl;
}

File: Lectures/Lecture 3/python/stack.py
from linked_list import LinkedList


class Stack(LinkedList):
    def __init__(self):
        super().__init__()

    def push(self, item):
        """ Add item to start of the linked list """
        self.prepend(item)

    def pop(self):
        """ Retrieve item from start of linked list, and remove it """
        data = self.head.data
        self.remove_first_node()
        return data

    def is_empty(self):
        # return self.head is None
        return len(self) == 0

    def peek(self):
        if self.head is None:
            return None

        return self.head.data

    def size(self):
        return len(self)


if __name__ == '__main__':
    my_stack = Stack()
    print("my_stack:", my_stack)

    print("Empty?", my_stack.is_empty())

    print()

    my_stack.push("hello")
    print("my_stack:", my_stack)
    my_stack.push("world")
    print("my_stack:", my_stack)
    print("Empty?", my_stack.is_empty())

    print("peek:", my_stack.peek())
    print()

    print("pop:", my_stack.pop())
    print("my_stack:", my_stack)
    print("pop:", my_stack.pop())
    print("my_stack:", my_stack)
    print("peek:", my_stack.peek())
